{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b5269f2f",
   "metadata": {},
   "source": [
    "### Python Programming Exercises 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f45a34d",
   "metadata": {},
   "source": [
    "#### 1.1. Given a triangle:\n",
    "If a and b are integers, calculate the integer which is the square of the hypotenuse\n",
    "So if a = 5 and b = 6, the answer = 61 (i.e. 52 + 62)\n",
    "Challenge: what is the answer if a = 10 and b = 12?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5be398f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "244\n"
     ]
    }
   ],
   "source": [
    "def hypotenuse(a,b):\n",
    "    return a**2 + b**2\n",
    "\n",
    "print(hypotenuse(10,12))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e93a456",
   "metadata": {},
   "source": [
    "#### 1.2. Given a string and two pairs of integers, return two slices of the string between the positions\n",
    "represented by the two number pairs inclusively.\n",
    "So given a string \"TheUniversityOfManchesterFacultyofBiologyMedicineAndHealth\" and number\n",
    "pairs 3,13 and 15,25 the output will be:\n",
    "University Manchester\n",
    "\n",
    "Challenge: what is the output given the following string and number pairs 137,143 and 274,280?\n",
    "GCTGAGACTTCCTGGACGGGGGACAGGCTGTGGGGTTTCTCAGATAACTGGGCCCCTGCGCTCAG\n",
    "GAGGCCTTCACCCTCTGCTCTGGGTAAAGTTCATTGGAACAGAAAGAAATGGATTTATCTGCTCT\n",
    "TCGCGTTGAAGAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCCCATCTGTC\n",
    "TGGAGTTGATCAAGGAACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTG\n",
    "AAACTTCTCAACCAGAAGAAAGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAAAG\n",
    "GAGCCTACAAGAAAGTACGAGATTTGAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2f53fe57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "University Manchester\n",
      "GAAGAA GAAGAA\n"
     ]
    }
   ],
   "source": [
    "example=\"TheUniversityOfManchesterFacultyofBiologyMedicineAndHealth\"\n",
    "challenge=\"GCTGAGACTTCCTGGACGGGGGACAGGCTGTGGGGTTTCTCAGATAACTGGGCCCCTGCGCTCAGGAGGCCTTCACCCTCTGCTCTGGGTAAAGTTCATTGGAACAGAAAGAAATGGATTTATCTGCTCTTCGCGTTGAAGAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCCCATCTGTCTGGAGTTGATCAAGGAACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTGAAACTTCTCAACCAGAAGAAAGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAAAGGAGCCTACAAGAAAGTACGAGATTTGAT\"\n",
    "\n",
    "def slice(string,pos1,pos2,pos3,pos4):\n",
    "    print(string[pos1:pos2]+\" \"+string[pos3:pos4])\n",
    "\n",
    "slice(example,3,13,15,25)\n",
    "slice(challenge,137,143,274,280)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e29ea305",
   "metadata": {},
   "source": [
    "#### 1.3. Given two positive integers a and b, return the sum of all odd integers from a to b inclusively.\n",
    "So if a = 50 and b = 100 the output = 1875\n",
    "Challenge: what is the output if a = 10 and b = 25?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "75161388",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "144\n",
      "144\n"
     ]
    }
   ],
   "source": [
    "#Becky\n",
    "def add_odds(a,b):\n",
    "    totalodds=0\n",
    "    for i in range(a,b+1):\n",
    "        if i%2!=0:\n",
    "            totalodds+=i\n",
    "    return totalodds\n",
    "\n",
    "def total_odds(a,b):\n",
    "    return sum([x for x in range(a,b+1) if x%2!=0])\n",
    "\n",
    "print(add_odds(10,25))\n",
    "print(total_odds(10,25))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bccfbe71",
   "metadata": {},
   "source": [
    "#### 1.4.Given a string representing a DNA sequence, print it in blocks, i.e. with gaps every so many bases.\n",
    "So given sequence:\n",
    "\n",
    "aggagtaagcccttgcaactggaaatacacccattg\n",
    "an output with a block size of 3 should look like:\n",
    "agg agt aag ccc ttg caa ctg gaa ata cac cca ttg\n",
    "\n",
    "Challenge: output the following sequence with a block size of 10:\n",
    "GCTGAGACTTCCTGGACGGGGGACAGGCTGTGGGGTTTCTCAGATAACTGGGCCCCTGCGCTCAG\n",
    "GAGGCCTTCACCCTCTGCTCTGGGTAAAGTTCATTGGAACAGAAAGAAATGGATTTATCTGCTCT\n",
    "TCGCGTTGAAGAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCCCATCTGTC\n",
    "TGGAGTTGATCAAGGAACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTG\n",
    "AAACTTCTCAACCAGAAGAAAGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAAAG\n",
    "GAGCCTACAAGAAAGTACGAGATTTGAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "241311da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "agg agt aag ccc ttg caa ctg gaa ata cac cca ttg\n",
      "GCTGAGACTT CCTGGACGGG GGACAGGCTG TGGGGTTTCT CAGATAACTG GGCCCCTGCG CTCAGGAGGC CTTCACCCTC TGCTCTGGGT AAAGTTCATT GGAACAGAAA GAAATGGATT TATCTGCTCT TCGCGTTGAA GAAGTACAAA ATGTCATTAA TGCTATGCAG AAAATCTTAG AGTGTCCCAT CTGTCTGGAG TTGATCAAGG AACCTGTCTC CACAAAGTGT GACCACATAT TTTGCAAATT TTGCATGCTG AAACTTCTCA ACCAGAAGAA AGGGCCTTCA CAGTGTCCTT TATGTAAGAA TGATATAACC AAAAGGAGCC TACAAGAAAG TACGAGATTT GAT\n",
      "\n",
      "aggagtaagcccttgcaactggaaatacacccattg\n",
      "agg agt aag ccc ttg caa ctg gaa ata cac cca ttg "
     ]
    }
   ],
   "source": [
    "sequence = \"aggagtaagcccttgcaactggaaatacacccattg\"\n",
    "challenge=\"GCTGAGACTTCCTGGACGGGGGACAGGCTGTGGGGTTTCTCAGATAACTGGGCCCCTGCGCTCAGGAGGCCTTCACCCTCTGCTCTGGGTAAAGTTCATTGGAACAGAAAGAAATGGATTTATCTGCTCTTCGCGTTGAAGAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCCCATCTGTCTGGAGTTGATCAAGGAACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTGAAACTTCTCAACCAGAAGAAAGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAAAGGAGCCTACAAGAAAGTACGAGATTTGAT\"\n",
    "\n",
    "def print_blocks(sequence, block_size):\n",
    "    print(*[sequence[index:index+block_size] for index in range(0,len(sequence),block_size)])\n",
    "\n",
    "print_blocks(sequence,3)\n",
    "print_blocks(challenge,10)\n",
    "print(\"\")\n",
    "\n",
    "dna=\"aggagtaagcccttgcaactggaaatacacccattg\"\n",
    "blocks_size=3\n",
    "\n",
    "for index, character in enumerate(dna):\n",
    "    print(character,end=\"\")\n",
    "print(\"\")\n",
    "\n",
    "for index, character in enumerate(dna):\n",
    "    print(character,end=\"\")\n",
    "    if (index+1) % blocks_size == 0:\n",
    "        print(\" \",end=\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f65f39db",
   "metadata": {},
   "source": [
    "#### 1.5.Write code to transcribe a DNA sequence. So given sequence:\n",
    "aggagtaagcccttgcaactggaaatacacccattg\n",
    "the output will be:\n",
    "aggaguaagcccuugcaacuggaaauacacccauug\n",
    "\n",
    "Challenge: transcribe the following:\n",
    "GCTGAGACTTCCTGGACGGGGGACAGGCTGTGGGGTTTCTCAGATAACTGGGCCCCTGCGCTCAGGAG\n",
    "GCCTTCACCCTCTGCTCTGGGTAAAGTTCATTGGAACAGAAAGAAATGGATTTATCTGCTCTTCGCGT\n",
    "TGAAGAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCCCATCTGTCTGGAGTTGA\n",
    "TCAAGGAACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTGAAACTTCTCAAC\n",
    "CAGAAGAAAGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAAAGGAGCCTACAAGAAAG\n",
    "TACGAGATTTGAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "236c324e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGGAGUAAGCCCUUGCAACUGGAAAUACACCCAUUG\n",
      "\n",
      "GCUGAGACUUCCUGGACGGGGGACAGGCUGUGGGGUUUCUCAGAUAACUGGGCCCCUGCGCUCAGGAGGCCUUCACCCUCUGCUCUGGGUAAAGUUCAUUGGAACAGAAAGAAAUGGAUUUAUCUGCUCUUCGCGUUGAAGAAGUACAAAAUGUCAUUAAUGCUAUGCAGAAAAUCUUAGAGUGUCCCAUCUGUCUGGAGUUGAUCAAGGAACCUGUCUCCACAAAGUGUGACCACAUAUUUUGCAAAUUUUGCAUGCUGAAACUUCUCAACCAGAAGAAAGGGCCUUCACAGUGUCCUUUAUGUAAGAAUGAUAUAACCAAAAGGAGCCUACAAGAAAGUACGAGAUUUGAU\n"
     ]
    }
   ],
   "source": [
    "DNA = \"aggagtaagcccttgcaactggaaatacacccattg\"\n",
    "challenge=\"GCTGAGACTTCCTGGACGGGGGACAGGCTGTGGGGTTTCTCAGATAACTGGGCCCCTGCGCTCAGGAGGCCTTCACCCTCTGCTCTGGGTAAAGTTCATTGGAACAGAAAGAAATGGATTTATCTGCTCTTCGCGTTGAAGAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCCCATCTGTCTGGAGTTGATCAAGGAACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTGAAACTTCTCAACCAGAAGAAAGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAAAGGAGCCTACAAGAAAGTACGAGATTTGAT\"\n",
    "\n",
    "def transcribe(seq):\n",
    "    return seq.upper().replace('T','U')\n",
    "\n",
    "print(transcribe(DNA))\n",
    "print(\"\")\n",
    "print(transcribe(challenge))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5a525cc",
   "metadata": {},
   "source": [
    "### Python Programming Exercises 2 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f087194e",
   "metadata": {},
   "source": [
    "### 2.1. Refer back to exercise 1.4, where we printed a DNA string in blocks, with a space between each block. Now further develop your code so that it displays a DNA string in the style used in GenBank records. \n",
    "So given the DNA sequence:\n",
    "\n",
    "GCTGAGACTTCCTGGACGGGGGACAGGCTGTGGGGTTTCTCAGATAACTGGGCCCCTGCGCTCAGGAGGC\n",
    "CTTCACCCTCTGCTCTGGGTAAAGTTCATTGGAACAGAAAGAAATGGATTTATCTGCTCTTCGCGTTGAA\n",
    "GAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCCCATCTGTCTGGAGTTGATCAAGG\n",
    "AACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTGAAACTTCTCAACCAGAAGAA\n",
    "AGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAAAGGAGCCTACAAGAAAGTACGAGATTT\n",
    "AGTCAACTTGTTGAAGAGCTATTGAAAATCATTTGTGCTTTTCAGCTTGACACAGGTTTGGAGTATGCAA\n",
    "ACAGCTATAATTTTGCAAAAAAGGAAAATAACTCTCCTGAACATCTAAAAGATGAAGTTTCTATCATCCA\n",
    "AAGTATGGGCTACAGAAACCGTGCCAAAAGACTTCTACAGAGTGAACCCGAAAATCCTTCCTTGCAGGAA\n",
    "ACCAGTCTCAGTGTCCAACTCTCTAACCTTGGAACTGTGAGAACTCTGAGGACAAAGCAGCGGATACAAC\n",
    "CTCAAAAGACGTCTGTCTACATTGAATTGGGATCTGATTCTTCTGAAGATACCGTTAATAAGGCAACTTA\n",
    "TTGCAGTGTGGGAGATCAAG\n",
    "\n",
    "The output should look as close as possible to:\n",
    "\n",
    "Hint: it’s a good idea to make your code into a function which has parameters for the block size and\n",
    "number of blocks per row, as well as the string to print. Also, see if you can ensure that the bases are\n",
    "always in lower case when printed regardless of the input."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9946a725",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 gctgagactt cctggacggg ggacaggctg tggggtttct cagataactg ggcccctgcg\n",
      "61 ctcaggaggc cttcaccctc tgctctgggt aaagttcatt ggaacagaaa gaaatggatt\n",
      "121 tatctgctct tcgcgttgaa gaagtacaaa atgtcattaa tgctatgcag aaaatcttag\n",
      "181 agtgtcccat ctgtctggag ttgatcaagg aacctgtctc cacaaagtgt gaccacatat\n",
      "241 tttgcaaatt ttgcatgctg aaacttctca accagaagaa agggccttca cagtgtcctt\n",
      "301 tatgtaagaa tgatataacc aaaaggagcc tacaagaaag tacgagattt agtcaacttg\n",
      "361 ttgaagagct attgaaaatc atttgtgctt ttcagcttga cacaggtttg gagtatgcaa\n",
      "421 acagctataa ttttgcaaaa aaggaaaata actctcctga acatctaaaa gatgaagttt\n",
      "481 ctatcatcca aagtatgggc tacagaaacc gtgccaaaag acttctacag agtgaacccg\n",
      "541 aaaatccttc cttgcaggaa accagtctca gtgtccaact ctctaacctt ggaactgtga\n",
      "601 gaactctgag gacaaagcag cggatacaac ctcaaaagac gtctgtctac attgaattgg\n",
      "661 gatctgattc ttctgaagat accgttaata aggcaactta ttgcagtgtg ggagatcaag\n"
     ]
    }
   ],
   "source": [
    "sequence = \"GCTGAGACTTCCTGGACGGGGGACAGGCTGTGGGGTTTCTCAGATAACTGGGCCCCTGCGCTCAGGAGGC CTTCACCCTCTGCTCTGGGTAAAGTTCATTGGAACAGAAAGAAATGGATTTATCTGCTCTTCGCGTTGAA GAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCCCATCTGTCTGGAGTTGATCAAGG AACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTGAAACTTCTCAACCAGAAGAA AGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAAAGGAGCCTACAAGAAAGTACGAGATTT AGTCAACTTGTTGAAGAGCTATTGAAAATCATTTGTGCTTTTCAGCTTGACACAGGTTTGGAGTATGCAA ACAGCTATAATTTTGCAAAAAAGGAAAATAACTCTCCTGAACATCTAAAAGATGAAGTTTCTATCATCCA AAGTATGGGCTACAGAAACCGTGCCAAAAGACTTCTACAGAGTGAACCCGAAAATCCTTCCTTGCAGGAA ACCAGTCTCAGTGTCCAACTCTCTAACCTTGGAACTGTGAGAACTCTGAGGACAAAGCAGCGGATACAAC CTCAAAAGACGTCTGTCTACATTGAATTGGGATCTGATTCTTCTGAAGATACCGTTAATAAGGCAACTTA TTGCAGTGTGGGAGATCAAG\"\n",
    "sequence=sequence.replace(\" \", \"\")\n",
    "\n",
    "def gbk_ify(seq, block_size, no_blocks):\n",
    "    #Make seq lower case\n",
    "    seq=seq.lower()\n",
    "    #Create empty lists\n",
    "    seqlist=[]\n",
    "    indexlist=[]\n",
    "    for index in range(0,len(seq),block_size):\n",
    "        #Append index of start of each block\n",
    "        indexlist.append(index)\n",
    "        #Append each block seq\n",
    "        seqlist.append(seq[index:index+block_size])\n",
    "    #Only include every nth index to print with beginning line, add 1\n",
    "    indexlist=[str(x+1) for x in indexlist[0::no_blocks]]\n",
    "    #Create empty list to zip two together\n",
    "    zippedlist=[]\n",
    "    #Insert no_blocks*elements of seqlist into indexlist after each element of index list\n",
    "    for i in range(len(indexlist)+1):\n",
    "        zippedlist+=indexlist[i:i+1]+seqlist[i*no_blocks:(i+1)*no_blocks]\n",
    "    #Unpack zipped list, print no_blocks+1 per line to include index\n",
    "    for element in range(0,len(zippedlist),no_blocks+1):\n",
    "        print(*zippedlist[element:element+no_blocks+1])\n",
    "\n",
    "gbk_ify(sequence,10,6)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ef86b9e",
   "metadata": {},
   "source": [
    "#### 2.2. Write a function to translate a DNA sequence into an amino acid sequence (don’t use imported modules to do this for now).\n",
    "You can find the standard genetic code table here\n",
    "https://www.genscript.com/tools/codon-table and elsewhere. Hint: one way is to create a dictionary\n",
    "to hold a translation table. Use single letter amino-acid codes, and assume coding starts from the\n",
    "first base only.\n",
    "\n",
    "So given a sequence:\n",
    "aggagtaagcccttgcaactggaaatacacccattg\n",
    "\n",
    "The output should look like:\n",
    "RSKPLQLEIHPL\n",
    "\n",
    "For bonus points, deal with any errors in the DNA string, e.g. incomplete codons at the end of the\n",
    "sequence, gaps or incorrect bases in the sequence. For now don’t consider ambiguity base codes like\n",
    "N.\n",
    "\n",
    "Challenge: now translate the following sequence:\n",
    "ATGGATTTATCTGCTCTTCGCGTTGAAGAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCC\n",
    "CATCTGTCTGGAGTTGATCAAGGAACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTGA\n",
    "AACTTCTCAACCAGAAGAAAGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "933bc6c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RSKPLQLEIHPL\n",
      "MDLSALRVEEVQNVINAMQKILECPICLELIKEPVSTKCDHIFCKFCMLKLLNQKKGPSQCPLCKNDITK\n"
     ]
    }
   ],
   "source": [
    "codon_table = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W',\n",
    "}\n",
    "\n",
    "example =\"aggagtaagcccttgcaactggaaatacacccattg\"\n",
    "challenge=\"ATGGATTTATCTGCTCTTCGCGTTGAAGAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCCCATCTGTCTGGAGTTGATCAAGGAACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTGAAACTTCTCAACCAGAAGAAAGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAA\"\n",
    "\n",
    "def amino_acid(string):\n",
    "    protein_seq=\"\"\n",
    "    string=string.upper()\n",
    "    for i in range(0,len(string),3):\n",
    "        if string[i:i+3] in codon_table:\n",
    "            protein_seq+=codon_table[string[i:i+3]]\n",
    "    return protein_seq\n",
    "\n",
    "print(amino_acid(example))\n",
    "print(amino_acid(challenge))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "414e0345",
   "metadata": {},
   "source": [
    "### 2.3. Write a function which generates the reverse complement of a sequence. \n",
    "Bonus points for dealing with gaps or incorrect base letters. So given a sequence:\n",
    "aggagtaagcccttgcaactggaaatacacccattg\n",
    "the output should look like:\n",
    "caatgggtgtatttccagttgcaagggcttactcct\n",
    "\n",
    "Challenge: output the reverse complement of:\n",
    "GCTGAGACTTCCTGGACGGGGGACAGGCTGTGGGGTTTCTCAGATAACTGGGCCCCTGCGCTCAGGAG\n",
    "GCCTTCACCCTCTGCTCTGGGTAAAGTTCATTGGAACAGAAAGAAATGGATTTATCTGCTCTTCGCGT\n",
    "TGAAGAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCCCATCTGTCTGGAGTTGA\n",
    "TCAAGGAACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTGAAACTTCTCAAC\n",
    "CAGAAGAAAGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAAAGGAGCCTACAAGAAAG\n",
    "TACGAGATTTGAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d224e228",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAATGGGTGTATTTCCAGTTGCAAGGGCTTACTCCT\n",
      "ATCAAATCTCGTACTTTCTTGTAGGCTCCTTTTGGTTATATCATTCTTACATAAAGGACACTGTGAAGGCCCTTTCTTCTGGTTGAGAAGTTTCAGCATGCAAAATTTGCAAAATATGTGGTCACACTTTGTGGAGACAGGTTCCTTGATCAACTCCAGACAGATGGGACACTCTAAGATTTTCTGCATAGCATTAATGACATTTTGTACTTCTTCAACGCGAAGAGCAGATAAATCCATTTCTTTCTGTTCCAATGAACTTTACCCAGAGCAGAGGGTGAAGGCCTCCTGAGCGCAGGGGCCCAGTTATCTGAGAAACCCCACAGCCTGTCCCCCGTCCAGGAAGTCTCAGC\n"
     ]
    }
   ],
   "source": [
    "example=\"aggagtaagcccttgcaactggaaatacacccattg\"\n",
    "challenge = \"GCTGAGACTTCCTGGACGGGGGACAGGCTGTGGGGTTTCTCAGATAACTGGGCCCCTGCGCTCAGGAG GCCTTCACCCTCTGCTCTGGGTAAAGTTCATTGGAACAGAAAGAAATGGATTTATCTGCTCTTCGCGT TGAAGAAGTACAAAATGTCATTAATGCTATGCAGAAAATCTTAGAGTGTCCCATCTGTCTGGAGTTGA TCAAGGAACCTGTCTCCACAAAGTGTGACCACATATTTTGCAAATTTTGCATGCTGAAACTTCTCAAC CAGAAGAAAGGGCCTTCACAGTGTCCTTTATGTAAGAATGATATAACCAAAAGGAGCCTACAAGAAAG TACGAGATTTGAT\"\n",
    "challenge=challenge.replace(\" \", \"\")\n",
    "\n",
    "def complement(seq):\n",
    "    complementary={\"A\":\"T\",\"T\":\"A\",\"C\":\"G\",\"G\":\"C\"}\n",
    "    reversedseq=seq[::-1].upper()\n",
    "    return reversedseq.translate(str.maketrans(complementary))\n",
    "\n",
    "print(complement(example))\n",
    "print(complement(challenge))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b80c5e35",
   "metadata": {},
   "source": [
    "#### 2.4. Combine translation and reverse complement functions to generate a six frame translation of a DNA sequence. \n",
    "\n",
    "This means you should translate three forward reading frames starting at the first,\n",
    "second and third base of the first codon of the forward sequence, and three reverse reading frames\n",
    "starting at the first, second and third base of the first codon of the reverse complement of the\n",
    "sequence. So given a sequence:\n",
    "aggagtaagcccttgcaactggaaatacacccattg\n",
    "\n",
    "The output should be something like:\n",
    "\n",
    "Forward\n",
    "\n",
    "1 RSKPLQLEIHPL\n",
    "\n",
    "2 GVSPCNWKYTH\n",
    "\n",
    "3 E*ALATGNTPI\n",
    "\n",
    "\n",
    "Reverse\n",
    "\n",
    "1 QWVYFQLQGLTP\n",
    "\n",
    "2 NGCISSCKGLL\n",
    "\n",
    "3 MGVFPVARAYS\n",
    "\n",
    "Challenge: print the six-frame translation of:\n",
    "GCTGAGACTTCCTGGACGGGGGACAGGCTGTGGGGTTTCTCAGATAACTGGGCCCCTGCGCTCAGGAGGCCT\n",
    "TCACCC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d6ac58b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward\n",
      "1 RSKPLQLEIHPL\n",
      "2 GVSPCNWKYTH\n",
      "3 E*ALATGNTPI\n",
      "Reverse\n",
      "1 QWVYFQLQGLTP\n",
      "2 NGCISSCKGLL\n",
      "3 MGVFPVARAYS\n",
      "\n",
      "Forward\n",
      "1 AETSWTGDRLWGFSDNWAPALRRPSP\n",
      "2 LRLPGRGTGCGVSQITGPLRSGGLH\n",
      "3 *DFLDGGQAVGFLR*LGPCAQEAFT\n",
      "Reverse\n",
      "1 G*RPPERRGPVI*ETPQPVPRPGSLS\n",
      "2 GEGLLSAGAQLSEKPHSLSPVQEVS\n",
      "3 VKAS*AQGPSYLRNPTACPPSRKSQ\n"
     ]
    }
   ],
   "source": [
    "example=\"aggagtaagcccttgcaactggaaatacacccattg\"\n",
    "challenge=\"GCTGAGACTTCCTGGACGGGGGACAGGCTGTGGGGTTTCTCAGATAACTGGGCCCCTGCGCTCAGGAGGCCTTCACCC\"\n",
    "\n",
    "codon_table = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W',\n",
    "}\n",
    "\n",
    "def complement(seq):\n",
    "    complementary={\"A\":\"T\",\"T\":\"A\",\"C\":\"G\",\"G\":\"C\"}\n",
    "    reversedseq=seq[::-1].upper()\n",
    "    return reversedseq.translate(str.maketrans(complementary))\n",
    "\n",
    "def amino_acid(string,n):\n",
    "    protein_seq=\"\"\n",
    "    string=string.upper()\n",
    "    for i in range(n,len(string),3):\n",
    "        if string[i:i+3] in codon_table:\n",
    "            protein_seq+=codon_table[string[i:i+3]]\n",
    "    return protein_seq\n",
    "\n",
    "print(\"Forward\")\n",
    "print(\"1\",amino_acid(example,0))\n",
    "print(\"2\",amino_acid(example,1))\n",
    "print(\"3\",amino_acid(example,2))\n",
    "\n",
    "print(\"Reverse\")\n",
    "print(\"1\",amino_acid(complement(example),0))\n",
    "print(\"2\",amino_acid(complement(example),1))\n",
    "print(\"3\",amino_acid(complement(example),2))\n",
    "print(\"\")\n",
    "print(\"Forward\")\n",
    "print(\"1\",amino_acid(challenge,0))\n",
    "print(\"2\",amino_acid(challenge,1))\n",
    "print(\"3\",amino_acid(challenge,2))\n",
    "\n",
    "print(\"Reverse\")\n",
    "print(\"1\",amino_acid(complement(challenge),0))\n",
    "print(\"2\",amino_acid(complement(challenge),1))\n",
    "print(\"3\",amino_acid(complement(challenge),2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28a0aae1",
   "metadata": {},
   "source": [
    "#### 2.5. Count single, di-nucleotide and tri-nucleotides in a sequence. \n",
    "So for sequence:\n",
    "\n",
    "aggagtaagcccttgcaactggaaatacacccattg\n",
    "\n",
    "The output would be something like:\n",
    "a 12\n",
    "g 8\n",
    "t 7\n",
    "c 9\n",
    "ag 1\n",
    "ga 1\n",
    "gt 1\n",
    "aa 3\n",
    "gc 2\n",
    "cc 2\n",
    "tt 1\n",
    "ct 1\n",
    "gg 1\n",
    "at 2\n",
    "ac 2\n",
    "tg 1\n",
    "agg 1\n",
    "agt 1\n",
    "aag 1\n",
    "ccc 1\n",
    "ttg 2\n",
    "caa 1\n",
    "ctg 1\n",
    "gaa 1\n",
    "ata 1\n",
    "cac 1\n",
    "cca 1\n",
    "for single bases for di-nucleotides for tri-nucleotides\n",
    "\n",
    "Hint – the above examples only count what is present so there are no counts of zero. For bonus\n",
    "points find all possible nucleotide combinations in advance then count those that are present. You\n",
    "can also provide warnings about incomplete groups of nucleotides at the end of the sequence, e.g. if\n",
    "there are two bases at the end when you are counting tri-nucleotides, and deal with gaps or nonstandard bases in the input sequence.\n",
    "\n",
    "Challenge: find the single, di-nucleotide and tri-nucleotide counts for:\n",
    "\n",
    "GAACCCGAAAATCCTTCCTTGCAGGAAACCAGTCTCAGTGTCCAACTCTCTAACCTTGGAACTGTGAGAA\n",
    "CTCTGAGGACAAAGCAGCGGATACAACCTCAAAAGACGTCTGTCTACATTGAATTGGGATCTGATTCTTC\n",
    "TGAAGATACCGTTAATAAGGCAACTTATTGCAGTGTGGGAGATCAAGAATTGTTACAAATCACCCCTCAA\n",
    "GGAACCAGGGATGAAATCAGTTTGGATTCTGCAAAAAAGGCTGCTTGTGAATTTTCTGAGACGGATGTAA\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a55ea0bb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Example:\n",
      "Single bases:\n",
      " {'a': 12, 'g': 7, 't': 7, 'c': 9}\n",
      "Dinucleotides:\n",
      " {'ag': 1, 'ga': 1, 'gt': 1, 'aa': 3, 'gc': 2, 'cc': 2, 'tt': 1, 'ct': 1, 'gg': 1, 'at': 2, 'ac': 2, 'tg': 1}\n",
      "Trinucleotides:\n",
      " {'agg': 1, 'agt': 1, 'aag': 1, 'ccc': 1, 'ttg': 2, 'caa': 1, 'ctg': 1, 'gaa': 1, 'ata': 1, 'cac': 1, 'cca': 1}\n",
      "\n",
      "Challenge\n",
      "Single bases:\n",
      " {'G': 62, 'A': 88, 'C': 58, 'T': 71}\n",
      "Dinucleotides:\n",
      " {'GA': 15, 'AC': 9, 'CC': 9, 'AA': 18, 'AT': 11, 'TT': 10, 'GC': 4, 'AG': 8, 'TC': 13, 'TG': 11, 'CA': 6, 'TA': 4, 'CT': 7, 'GG': 9, 'CG': 1, 'GT': 5}\n",
      "Trinucleotides:\n",
      " {'GAA': 8, 'CCC': 1, 'AAT': 2, 'CCT': 3, 'TCC': 1, 'TTG': 4, 'CAG': 2, 'ACC': 4, 'AGT': 4, 'CTC': 2, 'GTC': 2, 'CAA': 6, 'TCT': 7, 'AAC': 1, 'CTT': 1, 'GGA': 4, 'ACT': 3, 'GTG': 2, 'AGA': 1, 'CTG': 1, 'AGG': 2, 'ACA': 1, 'AAG': 4, 'CGG': 1, 'ATA': 1, 'ACG': 2, 'TAC': 1, 'ATT': 1, 'GAT': 6, 'GTT': 1, 'GCA': 2, 'TAT': 1, 'TGC': 1, 'TTA': 1, 'ATC': 2, 'AAA': 1, 'GCT': 2, 'TGT': 1, 'TTT': 1, 'GAG': 1, 'GTA': 1}\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "example=\"aggagtaagcccttgcaactggaaatacacccattg\"\n",
    "challenge=\"GAACCCGAAAATCCTTCCTTGCAGGAAACCAGTCTCAGTGTCCAACTCTCTAACCTTGGAACTGTGAGAA CTCTGAGGACAAAGCAGCGGATACAACCTCAAAAGACGTCTGTCTACATTGAATTGGGATCTGATTCTTC TGAAGATACCGTTAATAAGGCAACTTATTGCAGTGTGGGAGATCAAGAATTGTTACAAATCACCCCTCAA GGAACCAGGGATGAAATCAGTTTGGATTCTGCAAAAAAGGCTGCTTGTGAATTTTCTGAGACGGATGTAA\"\n",
    "challenge=challenge.replace(\" \",\"\")\n",
    "\n",
    "def count_chars(string,n):\n",
    "    nmers=[]\n",
    "    for i in range(0, len(string)-1,n):\n",
    "        nmers.append(string[i:i+n])\n",
    "    nmersdict={x:nmers.count(x) for x in nmers}\n",
    "    return nmersdict\n",
    "\n",
    "print(\"Example:\")\n",
    "print(\"Single bases:\\n\",count_chars(example,1))\n",
    "print(\"Dinucleotides:\\n\",count_chars(example,2))\n",
    "print(\"Trinucleotides:\\n\",count_chars(example,3))\n",
    "print(\"\")\n",
    "print(\"Challenge\")\n",
    "print(\"Single bases:\\n\", count_chars(challenge,1))\n",
    "print(\"Dinucleotides:\\n\",count_chars(challenge,2))\n",
    "print(\"Trinucleotides:\\n\",count_chars(challenge,3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa4ec001",
   "metadata": {},
   "source": [
    "#### 2.6. Develop a function which gives the GC content of a sequence. \n",
    "This is the number of G plus C bases in a sequence as a percentage of the total number of bases in the sequence. So for the\n",
    "sequence:\n",
    "aggagtaagcccttgcaactggaaatacacccattg\n",
    "\n",
    "the GC content is 47.22%.\n",
    "\n",
    "Hint: if you have completed 2.5 you can use that function to help achieve this.\n",
    "\n",
    "Challenge: what is the GC content of:\n",
    "GAACCCGAAAATCCTTCCTTGCAGGAAACCAGTCTCAGTGTCCAACTCTCTAACCTTGGAACTGTGAGAA\n",
    "CTCTGAGGACAAAGCAGCGGATACAACCTCAAAAGACGTCTGTCTACATTGAATTGGGATCTGATTCTTC\n",
    "TGAAGATACCGTTAATAAGGCAACTTATTGCAGTGTGGGAGATCAAGAATTGTTACAAATCACCCCTCAA\n",
    "GGAACCAGGGATGAAATCAGTTTGGATTCTGCAAAAAAGGCTGCTTGTGAATTTTCTGAGACGGATGTAA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "73943d70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47.22222222222222\n",
      "42.857142857142854\n"
     ]
    }
   ],
   "source": [
    "example=\"aggagtaagcccttgcaactggaaatacacccattg\"\n",
    "challenge=\"GAACCCGAAAATCCTTCCTTGCAGGAAACCAGTCTCAGTGTCCAACTCTCTAACCTTGGAACTGTGAGAA CTCTGAGGACAAAGCAGCGGATACAACCTCAAAAGACGTCTGTCTACATTGAATTGGGATCTGATTCTTC TGAAGATACCGTTAATAAGGCAACTTATTGCAGTGTGGGAGATCAAGAATTGTTACAAATCACCCCTCAA GGAACCAGGGATGAAATCAGTTTGGATTCTGCAAAAAAGGCTGCTTGTGAATTTTCTGAGACGGATGTAA\"\n",
    "#Remove whitespace from copy + pasted seq\n",
    "challenge=challenge.replace(\" \",\"\")\n",
    "\n",
    "#Make seq upper case\n",
    "#Count G + C, divide by length of string, *100\n",
    "def gc_content(seq):\n",
    "    seq=seq.upper()\n",
    "    return (float(seq.count(\"G\")+seq.count(\"C\"))/len(seq))*100\n",
    "\n",
    "print(gc_content(example))\n",
    "print(gc_content(challenge))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e742934",
   "metadata": {},
   "source": [
    "### Python programming exercises 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "22f9632e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aggagtaagcccttgcaactggaaatacacccattg\n",
      "\n",
      "agg agt aag ccc ttg caa ctg gaa ata cac cca ttg "
     ]
    }
   ],
   "source": [
    "dna=\"aggagtaagcccttgcaactggaaatacacccattg\"\n",
    "blocks_size=3\n",
    "\n",
    "#for each index of the character and each character as you iterate over DNA sequence\n",
    "for index, character in enumerate(dna):\n",
    "    #print the character, (on same line, not new line)\n",
    "    print(character,end=\"\")\n",
    "print(\"\\n\")\n",
    "\n",
    "#print each character, and if the index is divisible by the block size, print a space\n",
    "for index, character in enumerate(dna):\n",
    "    print(character,end=\"\")\n",
    "    if (index+1) % blocks_size == 0:\n",
    "        print(\" \",end=\"\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
